### PR TITLE
improve performance of controller

### LINKED ISSUE #753: Optimized performance of clusternet in large-scale CR scenarios.
#### What would you like to be added:
Optimized processing logic in clusternet-controller-manager

#### Why is this needed:

When there are more than 100,000 manifest or base resources in the cluster, the performance of controller-manager may be slow due to a large number of list operations. Optimization is needed for such scenarios.

The flame graph of pprof is shown below.
image

image

image

image

### PR DESCRIPTION
**Pull request by cumirror**
#### What type of PR is this?
This PR is aimed at optimizing the performance of controller components in large-scale scenarios.

When there are many cr objects, the efficiency of index is much higher than that of list by selector, as shown in this performance test.
**Here's the case**:
```
func BenchmarkGetByList(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if subs, err := subsLister.Subscriptions("clusternet-test").List(labels.SelectorFromSet(
			labels.Set{"6640f948-89af-4d1d-976d-08968895457d": "Subscription"})); err != nil || len(subs) != 1 {
			b.Fatalf("list fialed: %v %d", err, len(subs))
		}
	}
}

func BenchmarkGetByIndex(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if objs, err := sInformer.GetIndexer().ByIndex("uid", "6640f948-89af-4d1d-976d-08968895457d"); err != nil || len(objs) != 1 {
			b.Fatalf("list fialed: %v %d", err, len(objs))
		}
	}
}
```
**Benchmark results**:
```
2023-11-01T14:49:23+08:00: list subscription
2023-11-01T14:49:23+08:00: total count of subs is 100000
goos: darwin
goarch: arm64
BenchmarkGetByList-10                 46          24414396 ns/op
BenchmarkGetByIndex-10          15271741                79.20 ns/op
```
When searching from a list of 100,000 cr objects, the efficiency of index is much higher than that of list.

#### What this PR does / why we need it:
This PR focuses on optimizing the base lookup process in the controller, which involves helmcharts, manifests, and subs processing.

#### Which issue(s) this PR fixes:

Fixes #753

### CODE REVIEW

File: pkg/utils/feed.go [hunk 1]
```diff
@@ -136,17 +137,19 @@ func FindObsoletedFeeds(oldFeeds []appsapi.Feed, newFeeds []appsapi.Feed) []apps
...
-		bases, err := baseLister.List(labels.SelectorFromSet(labels.Set{
-			uid: "Base",
-		}))
-		if err != nil {
+		if objs, err := baseIndexer.ByIndex("uid", uid); err != nil {
 			klog.ErrorDepth(5, err)
 			continue
+		} else {
+			if len(objs) == 1 {
+				allBases = append(allBases, objs[0].(*appsapi.Base))
```
Comment (dixudx) [inline pkg/utils/feed.go:148]: The length of `objs` could only be 0 or 1. Right?

**Review by dixudx — COMMENTED** [review]
We're getting close.

File: pkg/controllers/apps/base/base.go [hunk 2]
```diff
@@ -96,8 +98,17 @@ func NewController(
 			return true, nil
 		})
 
+	err := baseInformer.Informer().AddIndexers(cache.Indexers{"uid": utils.BaseUidIndexFunc})
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go]: done

File: pkg/controllers/apps/base/base.go [hunk 3]
```diff
@@ -255,3 +266,30 @@ func (c *Controller) patchBaseLabels(base *appsapi.Base, labels map[string]*stri
 		patchData,
 		metav1.PatchOptions{})
 }
+
+func (c *Controller) FindBaseByUID(uid string) (*appsapi.Base, error) {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go:270]: done

[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go:270]: done

File: pkg/controllers/apps/base/base.go [hunk 4]
```diff
@@ -255,3 +266,30 @@ func (c *Controller) patchBaseLabels(base *appsapi.Base, labels map[string]*stri
...
+	if objs, err := c.baseIndexer.ByIndex("uid", uid); err != nil {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go]: done

File: pkg/controllers/apps/base/base.go [hunk 5]
```diff
@@ -255,3 +266,30 @@ func (c *Controller) patchBaseLabels(base *appsapi.Base, labels map[string]*stri
...
+	} else {
+		if len(objs) == 1 {
+			return objs[0].(*appsapi.Base), nil
+		} else {
+			return nil, fmt.Errorf("find base by uid %s failed", uid)
+		}
+	}
+}
+
+func (c *Controller) FindBaseBySubUID(subUid string) ([]*appsapi.Base, error) {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go:279]: done

File: pkg/controllers/apps/base/base.go [hunk 6]
```diff
@@ -255,3 +266,30 @@ func (c *Controller) patchBaseLabels(base *appsapi.Base, labels map[string]*stri
...
+	var bases []*appsapi.Base
+	if objs, err := c.baseIndexer.ByIndex("subUid", subUid); err != nil {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/controllers/apps/base/base.go]: ok, done.

File: pkg/utils/deployer.go [hunk 7]
```diff
@@ -933,3 +933,21 @@ func UpdateDescriptionStatus(desc *appsapi.Description, status *appsapi.Descript
 		return err2
 	})
 }
+
+func BaseUidIndexFunc(obj interface{}) ([]string, error) {
+	base := obj.(*appsapi.Base)
+	return []string{string(base.UID)}, nil
+}
+
+func BaseSubUidIndexFunc(obj interface{}) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/utils/deployer.go:945]: done

File: pkg/utils/feed.go [hunk 8]
```diff
@@ -136,17 +137,19 @@ func FindObsoletedFeeds(oldFeeds []appsapi.Feed, newFeeds []appsapi.Feed) []apps
...
 	allBases := []*appsapi.Base{}
 	for _, uid := range uids {
-		bases, err := baseLister.List(labels.SelectorFromSet(labels.Set{
-			uid: "Base",
-		}))
-		if err != nil {
+		if objs, err := baseIndexer.ByIndex("uid", uid); err != nil {
 			klog.ErrorDepth(5, err)
 			continue
+		} else {
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/utils/feed.go:146]: done

File: pkg/utils/feed.go [hunk 9]
```diff
@@ -136,17 +137,19 @@ func FindObsoletedFeeds(oldFeeds []appsapi.Feed, newFeeds []appsapi.Feed) []apps
...
+			if len(objs) == 1 {
+				allBases = append(allBases, objs[0].(*appsapi.Base))
+			} else {
+				klog.ErrorDepth(5, fmt.Errorf("find base by uid %s failed", uid))
```
[the remark this answers is not in the document]
Comment (cumirror) [reply pkg/utils/feed.go]: done

File: pkg/utils/feed.go (hunk 1 above)
Comment (cumirror) [reply pkg/utils/feed.go]: yes, find by base uid so result is unique.

File: pkg/utils/deployer.go [hunk 10]
```diff
@@ -933,3 +933,32 @@ func UpdateDescriptionStatus(desc *appsapi.Description, status *appsapi.Descript
...
+	base, ok := obj.(*appsapi.Base)
+	if !ok {
+		return nil, fmt.Errorf("object is not a Base %#v", obj)
+	}
+
+	subUid, ok := base.Labels[known.ConfigSubscriptionUIDLabel]
+	if ok {
+		return []string{subUid}, nil
+	}
+	return []string{""}, nil
```
Comment (dixudx) [inline pkg/utils/deployer.go:955]: This should return non-nil error, instead of nil.

Comment (cumirror) [reply pkg/utils/deployer.go]: done

**Pull request merged**